test/test_common/file_system_for_test.h

```diff
@@ -2,10 +2,127 @@
 
 #include "envoy/filesystem/filesystem.h"
 
+#include "common/filesystem/file_shared_impl.h"
+
+#include "absl/synchronization/mutex.h"
+
 namespace Envoy {
 
 namespace Filesystem {
-Instance& fileSystemForTest();
+
+struct MemFileInfo {
+  absl::Mutex lock_;
+  std::string data_ GUARDED_BY(lock_);
+};
+
+class MemfileImpl : public FileSharedImpl {
+public:
+  MemfileImpl(const std::string& path, std::shared_ptr<MemFileInfo>& info)
+      : FileSharedImpl(path), info_(info) {}
+
+protected:
+  Api::IoCallBoolResult open(FlagSet flag) override {
+    ASSERT(!isOpen());
+    flags_ = flag;
+    fd_++;
+    return resultSuccess(true);
+  }
+
+  Api::IoCallSizeResult write(absl::string_view buffer) override {
+    absl::MutexLock l(&info_->lock_);
+    if (!flags_.test(File::Operation::Append)) {
+      info_->data_.clear();
+    }
+    info_->data_.append(buffer);
+    const ssize_t size = info_->data_.size();
+    return resultSuccess(size);
+  }
+
+  Api::IoCallBoolResult close() override {
+    ASSERT(isOpen());
+    fd_--;
+    return resultSuccess(true);
+  }
+
+private:
+  FlagSet flags_;
+  std::shared_ptr<MemFileInfo> info_;
+};
+
+class MemfileInstanceImpl : public Instance {
+public:
+  MemfileInstanceImpl();
+
+  FilePtr createFile(const std::string& path) override {
+    absl::MutexLock m(&lock_);
+    if (file_system_->fileExists(path) || !use_memfiles_) {
+      return file_system_->createFile(path);
+    }
+    auto it = files_.find(path);
+    if (it == files_.end()) {
+      it = files_.emplace(path, std::make_shared<MemFileInfo>()).first;
+    }
+    std::shared_ptr<MemFileInfo> info = it->second;
+    return FilePtr{new MemfileImpl(path, info)};
+  }
+
+  bool fileExists(const std::string& path) override {
+    absl::MutexLock m(&lock_);
+    auto it = files_.find(path);
+    return (it != files_.end() || file_system_->fileExists(path));
```

Comment:
Why do we resort to hitting the real file system? Can we go all in on MemFiles at some point? Or is this needed to avoid having this PR pull all the files in testdata/ into the test .cc files?

Reply:
yeah exactly, many tests use testdata which is still in files, and given some is used for the hot restart tests I think it'll need to stay that way